xprocess.py

```diff
@@ -203,37 +203,60 @@ class ProcessStarter:
     before presuming the attached process dead.
     """
 
+    line_wait_timeout = 120
+    """The maximum time ProcessStarter.wait will hang waiting for a new
+    line when trying to match pattern before raising TimeoutError"""
+
     def __init__(self, control_dir, process):
         self.control_dir = control_dir
         self.process = process
 
     @abc.abstractproperty
     def args(self):
         "The args to start the process"
+        raise NotImplementedError
```

Comment:
Curious, why was this added? Doesn't the `abc.abstractproperty` decorator accomplishes the same?

Reply:
As the code was before it would just return `None` if you try fetching the `args` property when it's not implemented/defined. No exceptions would be raised and execution would proceed normaly. I prefer following "Explicit is better than implicit" and if someone tries to fetch and use a property that should be implemented but is not, raise a comprehensive exception and interrupt execution.

Reply:
To illustrate:
```python
import abc

class Parent:
    @abc.abstractproperty
    def myprop(self):
        "my test property"

class MyClass(Parent):
    pass

myclass = MyClass()

print(f"My prop value: {myclass.myprop}")
```
> My prop value: None

Opposed to:
```python
import abc

class Parent:
    @abc.abstractproperty
    def myprop(self):
        "my test property"
        raise NotImplementedError

class MyClass(Parent):
    pass

myclass = MyClass()

print(f"My prop value: {myclass.myprop}")
```

> Traceback (most recent call last):
>   File "test.py", line 15, in <module>
>     print(f"My prop value: {myclass.myprop}")
>   File "test.py", line 8, in myprop
>     raise NotImplementedError
> NotImplementedError

Reply:
My question was that `abc.abstractproperty` should raise `NotImplementedError`, but now I see the issue is that the class doesn't subclass from `abc.ABC` (which is responsible for raising the exception really):

```python
from abc import ABC, abstractproperty

class C(ABC):

    @abstractproperty
    def name(self): ...

c = C()
```

In this case, you can't even instantiate `C`:

```
Traceback (most recent call last):
  File "foo.py", line 10, in <module>
    c = C()
TypeError: Can't instantiate abstract class C with abstract methods name
```

So using `@abc.abstractproperty` without subclassing `abc.ABC` seems like an error here.

Reply:
Nice catch! I'm considering releasing a patch to fix this, 0.16.1?

Reply:
I just saw that @abc.abstractproperty has been

> Deprecated since version 3.3: It is now possible to use property, property.getter(), property.setter() and property.deleter() with abstractmethod(), making this decorator redundant

so on second thought, I think we could instead update the code and not use `@abc.abstractproperty` at all.

Reply:
>  I think we could instead update the code and not use @abc.abstractproperty at all.

Agreed! 

Btw, when I said "... seems like an error here" I didn't mean that the code _as it stands now_ was incorrect, just that whoever added `@abstractproperty` incorrectly assumed it would raise `NotImplementedError`, which was not the case until you fixed it by raising it explicitly.

In summary, I agree that no new release is necessary, we should just remove the decorator for sanity's sake (and also because it is deprecated as you mentioned).

Reply:
Okay, I'll write a PR addressing this in a few days. 👍🏻